tests/xharness/TestReporter.cs

```diff
@@ -0,0 +1,490 @@
+					return true;
+				}
+			}
+			return false;
+		}
+
+		// kill any process 
+		Task KillAppProcess (int pid, CancellationTokenSource cancellationSource) { 
+				var launchTimedout = cancellationSource.IsCancellationRequested;
+				var timeoutType = launchTimedout ? "Launch" : "Completion";
+				var timeoutValue = launchTimedout ? runner.LaunchTimeout : Timeout.TotalSeconds;
+
+				runner.MainLog.WriteLine ($"{timeoutType} timed out after {timeoutValue} seconds");
+				return ProcessManager.KillTreeAsync (pid, runner.MainLog, true);
+		}
+
+		async Task CollectResult (Task<ProcessExecutionResult> processExecution)
+		{ 
+			// wait for the execution of the process, once that is done, perform all the parsing operations and
+			// leave a clean API to be used by AppRunner, hidding all the diff details
+			var result = await processExecution;
+			if (!waitedForExit && !result.TimedOut) {
+				// mlaunch couldn't wait for exit for some reason. Let's assume the app exits when the test listener completes.
+				runner.MainLog.WriteLine ("Waiting for listener to complete, since mlaunch won't tell.");
+				if (!await listener.CompletionTask.TimeoutAfter (Timeout - TimeoutWatch.Elapsed)) {
+					result.TimedOut = true;
+				}
+			}
+
+			if (result.TimedOut) {
+				timedout = true;
+				Success = false;
+				runner.MainLog.WriteLine (timeoutMessage, Timeout.TotalMinutes);
+			} else if (result.Succeeded) {
+				runner.MainLog.WriteLine (completionMessage);
+				Success = true;
+			} else {
+				runner.MainLog.WriteLine (failureMessage);
+				Success = false;
+			}
+		}
+
+		public void LaunchCallback (Task<bool> launchResult)
+		{
+			if (launchResult.IsFaulted) {
+				runner.MainLog.WriteLine ("Test launch failed: {0}", launchResult.Exception);
+			} else if (launchResult.IsCanceled) {
+				runner.MainLog.WriteLine ("Test launch was cancelled.");
+			} else if (launchResult.Result) {
+				runner.MainLog.WriteLine ("Test run started");
+			} else {
+				cancellationTokenSource.Cancel ();
+				runner.MainLog.WriteLine ("Test launch timed out after {0} minute(s).", runner.LaunchTimeout);
+				timedout = true;
+			}
+		}
+
+		public async Task CollectSimulatorResult (Task<ProcessExecutionResult> processExecution)
+		{
+			isSimulatorTest = true;
+			await CollectResult (processExecution);
+
+			if (!Success.Value) {
+				var (pid, launchFailure) = await GetPidFromRunLog ();
+				this.launchFailure = launchFailure;
+				if (pid > 0) {
+					await KillAppProcess (pid, cancellationTokenSource);
+				} else {
+					runner.MainLog.WriteLine ("Could not find pid in mtouch output.");
+				}
+			}
+		}
+
+		public async Task CollectDeviceResult (Task<ProcessExecutionResult> processExecution)
+		{
+			isSimulatorTest = false;
+			await CollectResult (processExecution);
+		}
+
+		async Task<(string ResultLine, bool Failed)> GetResultLine (string logPath)
+		{
+			string resultLine = null;
```

Comment:
Same logic we had, just got moved to a method and made async since we could do it without blocking.